narwhals/_pandas_like/namespace.py

```diff
@@ -250,3 +251,130 @@ def concat(
                 backend_version=self._backend_version,
             )
         raise NotImplementedError
+
+    def when(
+        self,
+        *predicates: IntoPandasLikeExpr | Iterable[IntoPandasLikeExpr],
+    ) -> PandasWhen:
+        plx = self.__class__(self._implementation, self._backend_version)
+        if predicates:
+            condition = plx.all_horizontal(*flatten(predicates))
+        else:
+            msg = "at least one predicate needs to be provided"
+            raise TypeError(msg)
+
+        return PandasWhen(condition, self._implementation, self._backend_version)
+
+
+def _when_then_value_arg_process(
+    plx: PandasLikeNamespace,
+    value: PandasLikeExpr | PandasLikeSeries | Any,
+    *,
+    shape: tuple[int] | None = None,
+    series_with_shape: PandasLikeSeries | None = None,
+) -> PandasLikeExpr:
+    from narwhals.dependencies import get_numpy
+
+    np = get_numpy()
+
+    if not np:
+        msg = "numpy is required for this function"
+        raise ImportError(msg)
+    if isinstance(value, PandasLikeExpr):
+        return value
+    elif isinstance(value, PandasLikeSeries):
+        return plx._create_expr_from_series(value)
+    elif isinstance(value, np.ndarray):
```

Comment:
Interestingly enough `polars` considers python arrays as scalars for `when-then-otherwise`, but numpy arrays and series as iterables to zip with

Reply:
Also note that the pandas code does not guarantee type safety

@aivanoved TODO for me

Reply:
This is out of the scope of this pr as seems to have to do with `PandasLikeSeries.zip_with`